File: common/java/src/main/java/org/aiddl/common/planning/state_variable/ForwardSearchPlanIterator.java
package org.aiddl.common.planning.state_variable;

import java.util.Map;
import org.aiddl.common.CommonTerm;
import org.aiddl.common.planning.PlanningTerm;
import org.aiddl.common.search.GraphSearch;
import org.aiddl.core.function.FunctionRegistry;
import org.aiddl.core.interfaces.ConfigurableFunction;
import org.aiddl.core.interfaces.Function;
import org.aiddl.core.interfaces.InitializableFunction;
import org.aiddl.core.representation.SetTerm;
import org.aiddl.core.representation.SymbolicTerm;
import org.aiddl.core.representation.Term;
import org.aiddl.core.tools.Logger;

public class ForwardSearchPlanIterator implements InitializableFunction, ConfigurableFunction {
	
	FunctionRegistry fReg;
	GraphSearch search;
	
	SymbolicTerm heuristicName = Term.sym("causal-graph");
	
	String name = "PlanIterator";
	int verbose = 0;

	@Override
	public void configure(Map<Term, Term> settings, FunctionRegistry fReg) {
		this.fReg = fReg;
		this.verbose = settings.getOrDefault(Term.sym("verbose"), Term.integer(0)).getIntValue();
		this.heuristicName = settings.getOrDefault(PlanningTerm.Heuristic, this.heuristicName).asSym();
	}
	
	@Override
	public void initialize(Term problem) {
		SetTerm O = (SetTerm) problem.get(PlanningTerm.Operators);
		SetTerm s0 = problem.get(PlanningTerm.InitialState).asSet();
		SetTerm g = problem.get(PlanningTerm.Goal).asSet();
		
//		LockablSet<Operator> A = new LinkedHashSet<>();
//		for ( Term o : O ) {
//			A.add(new Operator((TupleTerm) o));
//		}
		if ( verbose >= 1 ) {
			Logger.msg(name, "Initial State: " + s0);
			Logger.msg(name, "Goal         : " + g);
		}
	
		GoalTest goalTest = new GoalTest();
		
		goalTest.initialize(g);

		OperatorReachableEnumerator groundOps = new OperatorReachableEnumerator();
		
		Term O_ground = groundOps.apply(problem);
		
//		System.out.println("Applicable ops: " + O_ground.size());
		
		Function h = HeuristicFactory.createHeuristic(this.heuristicName, problem, O_ground.asSet());
		
		Expand expand = new Expand();
		expand.initialize(O_ground);
				
		search = new GraphSearch();
		search.setExpansion(expand);
		search.setGoalTest(goalTest);
		search.setHeuristic(h);
		search.setSearchParameters(Term.rational(5, 10), true);
		search.initialize(Term.list(s0));
	}
	
	Term last_solution;
	boolean done = false;
	
	@Override
	public Term apply(Term args) {
		if ( args.equals(Term.sym("next")) ) {
			if ( done ) {
				return CommonTerm.NIL;
			}
			Term next_goal_node = search.apply(Term.tuple(GraphSearch.Search));
			if ( next_goal_node.equals(CommonTerm.NIL) ) {
				done = true;
				return next_goal_node;
			}
//			System.out.println(next_goal_node);
			Term path = search.apply(Term.tuple(GraphSearch.Get, GraphSearch.Path, next_goal_node));
			this.last_solution = path;
			
			return path;
		}
		return CommonTerm.NIL;
	}
}

File: common/scala/src/main/scala/org/aiddl/common/scala/planning/state_variable/ForwardSearchPlanIterator.scala
package org.aiddl.common.scala.planning.state_variable

import org.aiddl.core.scala.function.Initializable
import org.aiddl.core.scala.representation.*
import org.aiddl.common.scala.search.TermGraphSearch
import org.aiddl.common.scala.planning.state_variable.heuristic.{CausalGraphHeuristic, FastForwardHeuristic, Heuristic, SumCostHeuristic}
import org.aiddl.common.scala.planning.PlanningTerm.*

object ForwardSearchPlanIterator {
    def apply(): ForwardSearchPlanIterator ={
        new ForwardSearchPlanIterator(List((new CausalGraphHeuristic, Num(1))))
    }
}

class ForwardSearchPlanIterator(hs: Seq[(Heuristic, Num)]) extends TermGraphSearch {
    //val f_h = new CausalGraphHeuristic
    val f_goal = new GoalTest
    val f_exp = new Expansion

    hs.foreach((h, o) => this.addHeuristic(o, h))


    var groundOperators: Term = _
    var needGrounding: Boolean = true

    override def init( p: Term ) = {
        var problem = p
        if needGrounding then {
            val ground = new ReachableOperatorEnumerator
            groundOperators = ground(problem)
            problem = problem.asCol.put(KeyVal(Operators, groundOperators))
        } else {
            groundOperators = problem(Operators)
        }

        f_exp.init(groundOperators)
        this.heuristics.foreach(h =>
            if  h.isInstanceOf[Initializable]
            then h.asInstanceOf[Initializable].init(problem))
        //f_h.init(problem)
        f_goal.init(p(Goal))
        super.init(ListTerm(p(InitialState)))
    }

    override def isGoal(node: Term ): Boolean = f_goal(node).asBool.v
    override def expand(node: Term ): Seq[(Term, Term)] = f_exp.expand(node)
}
